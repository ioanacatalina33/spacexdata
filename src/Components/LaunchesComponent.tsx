import { Launch } from 'store/types/spaceXTypes';
import LaunchComponent from './LaunchComponent';
import { useMemo, useState } from 'react';

export default function LaunchesComponent({
  launches,
}: {
  launches: Launch[];
}) {
  const [showOnlyWithShips, setShowOnlyWithShips] = useState(false);

  const filteredLaunches = useMemo(
    () =>
      showOnlyWithShips
        ? launches.filter((launch) => !!launch.ships.length)
        : launches,
    [launches, showOnlyWithShips],
  );
  return (
    <div>
      <h1>Launches</h1>
      <button
        style={{ minWidth: '15rem' }}
        onClick={() => setShowOnlyWithShips((prev) => !prev)}
      >
        {showOnlyWithShips ? 'Show all' : 'Show only with Ships'}
      </button>
      <h4>Showing: {filteredLaunches.length} launches</h4>
      <table>
        {filteredLaunches.map((launch, i) => (
          <tr
            key={launch.id}
            style={{ backgroundColor: i % 2 ? '#cceeff' : '#d7ffcc' }}
          >
            <LaunchComponent launch={launch} />
          </tr>
        ))}
      </table>
    </div>
  );
}
