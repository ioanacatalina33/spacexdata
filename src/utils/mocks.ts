import { Launch, Ship } from 'store/types/spaceXTypes';

export const shipMock: Ship = {
  id: '123',
  model: 'ModelX',
  name: 'ShipX',
  yearBuild: '2022',
  type: 'typeX',
};

export const mockLaunches: Launch[] = [
  {
    name: 'Launch1',
    id: '1',
    details: 'details 1',
    success: false,
    rocketId: 'rocket1',
    ships: [
      {
        id: '1',
        model: 'Model1',
        name: 'ship1',
        type: 'typee1',
        yearBuild: '2022',
      },
    ],
  },
  {
    name: 'Launch2',
    id: '2',
    details: 'details 2',
    success: true,
    rocketId: 'rocket2',
    ships: [],
  },
  {
    name: 'Launch3',
    id: '3',
    details: 'details 3',
    success: true,
    rocketId: 'rocket3',
    ships: [],
  },
  {
    name: 'Launch4',
    id: '4',
    details: 'details 4',
    success: true,
    rocketId: 'rocket4',
    ships: [
      {
        id: '2',
        model: 'Model2',
        name: 'ship2',
        type: 'typee2',
        yearBuild: '2022',
      },
    ],
  },
];
