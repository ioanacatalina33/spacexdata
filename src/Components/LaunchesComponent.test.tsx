import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { mockLaunches } from '../utils/mocks';
import LaunchesComponent from './LaunchesComponent';
import userEvent from '@testing-library/user-event';

describe('LaunchesComponent tests', () => {
  it('should render correctly', () => {
    const shipComponent = render(<LaunchesComponent launches={mockLaunches} />);
    expect(shipComponent).toMatchSnapshot();
  });
});

it('should show 4 launches', () => {
  const shipComponent = render(<LaunchesComponent launches={mockLaunches} />);
  expect(shipComponent.getByText('Showing: 4 launches')).toBeInTheDocument();
});

it('should filter', async () => {
  const shipComponent = render(<LaunchesComponent launches={mockLaunches} />);
  expect(shipComponent.getByText('Showing: 4 launches')).toBeInTheDocument();
  const button = screen.getByRole('button', { name: 'Show only with Ships' });

  expect(button).toBeInTheDocument();

  userEvent.click(button);

  const component = await shipComponent.findByText('Showing: 2 launches');
  expect(component).toBeInTheDocument();
  expect(
    shipComponent.queryByText('Showing: 4 launches'),
  ).not.toBeInTheDocument();
});
