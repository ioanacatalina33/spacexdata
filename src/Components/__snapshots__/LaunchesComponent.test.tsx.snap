// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchesComponent tests should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          Launches
        </h1>
        <button
          style="min-width: 15rem;"
        >
          Show only with Ships
        </button>
        <h4>
          Showing: 
          4
           launches
        </h4>
        <table>
          <tr
            style="background-color: rgb(215, 255, 204);"
          >
            <td>
              <h3>
                Name: 
                Launch1
              </h3>
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              success: 
              no
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              details 1
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              rocketId: 
              rocket1
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              Ships: 
              <div
                style="display: flex; flex-direction: column;"
              >
                <ul
                  style="text-align: left;"
                >
                  <li>
                    Name: 
                    ship1
                  </li>
                  <li>
                     Model: 
                    Model1
                  </li>
                  <li>
                    Type: 
                    typee1
                  </li>
                  <li>
                     Year: 
                    2022
                  </li>
                </ul>
              </div>
            </td>
          </tr>
          <tr
            style="background-color: rgb(204, 238, 255);"
          >
            <td>
              <h3>
                Name: 
                Launch2
              </h3>
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              success: 
              yes
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              details 2
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              rocketId: 
              rocket2
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              Ships: 
              <div
                style="display: flex; flex-direction: column;"
              />
            </td>
          </tr>
          <tr
            style="background-color: rgb(215, 255, 204);"
          >
            <td>
              <h3>
                Name: 
                Launch3
              </h3>
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              success: 
              yes
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              details 3
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              rocketId: 
              rocket3
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              Ships: 
              <div
                style="display: flex; flex-direction: column;"
              />
            </td>
          </tr>
          <tr
            style="background-color: rgb(204, 238, 255);"
          >
            <td>
              <h3>
                Name: 
                Launch4
              </h3>
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              success: 
              yes
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              details 4
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              rocketId: 
              rocket4
            </td>
            <td
              style="min-width: 10rem; padding: 1rem;"
            >
              Ships: 
              <div
                style="display: flex; flex-direction: column;"
              >
                <ul
                  style="text-align: left;"
                >
                  <li>
                    Name: 
                    ship2
                  </li>
                  <li>
                     Model: 
                    Model2
                  </li>
                  <li>
                    Type: 
                    typee2
                  </li>
                  <li>
                     Year: 
                    2022
                  </li>
                </ul>
              </div>
            </td>
          </tr>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Launches
      </h1>
      <button
        style="min-width: 15rem;"
      >
        Show only with Ships
      </button>
      <h4>
        Showing: 
        4
         launches
      </h4>
      <table>
        <tr
          style="background-color: rgb(215, 255, 204);"
        >
          <td>
            <h3>
              Name: 
              Launch1
            </h3>
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            success: 
            no
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            details 1
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            rocketId: 
            rocket1
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            Ships: 
            <div
              style="display: flex; flex-direction: column;"
            >
              <ul
                style="text-align: left;"
              >
                <li>
                  Name: 
                  ship1
                </li>
                <li>
                   Model: 
                  Model1
                </li>
                <li>
                  Type: 
                  typee1
                </li>
                <li>
                   Year: 
                  2022
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr
          style="background-color: rgb(204, 238, 255);"
        >
          <td>
            <h3>
              Name: 
              Launch2
            </h3>
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            success: 
            yes
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            details 2
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            rocketId: 
            rocket2
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            Ships: 
            <div
              style="display: flex; flex-direction: column;"
            />
          </td>
        </tr>
        <tr
          style="background-color: rgb(215, 255, 204);"
        >
          <td>
            <h3>
              Name: 
              Launch3
            </h3>
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            success: 
            yes
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            details 3
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            rocketId: 
            rocket3
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            Ships: 
            <div
              style="display: flex; flex-direction: column;"
            />
          </td>
        </tr>
        <tr
          style="background-color: rgb(204, 238, 255);"
        >
          <td>
            <h3>
              Name: 
              Launch4
            </h3>
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            success: 
            yes
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            details 4
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            rocketId: 
            rocket4
          </td>
          <td
            style="min-width: 10rem; padding: 1rem;"
          >
            Ships: 
            <div
              style="display: flex; flex-direction: column;"
            >
              <ul
                style="text-align: left;"
              >
                <li>
                  Name: 
                  ship2
                </li>
                <li>
                   Model: 
                  Model2
                </li>
                <li>
                  Type: 
                  typee2
                </li>
                <li>
                   Year: 
                  2022
                </li>
              </ul>
            </div>
          </td>
        </tr>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
